src/compute-types/src/plan/flat_plan.rs

```diff
@@ -0,0 +1,1156 @@
+                        threshold_plan,
+                    };
+                    insert_node(node_id, node);
+
+                    todo.push(*input);
+                }
+                Plan::Union {
+                    inputs,
+                    consolidate_output,
+                    node_id,
+                } => {
+                    let node = Union {
+                        inputs: inputs.iter().map(|i| i.node_id()).collect(),
+                        consolidate_output,
+                    };
+                    insert_node(node_id, node);
+
+                    todo.extend(inputs.into_iter());
+                }
+                Plan::ArrangeBy {
+                    input,
+                    forms,
+                    input_key,
+                    input_mfp,
+                    node_id,
+                } => {
+                    let node = ArrangeBy {
+                        input: input.node_id(),
+                        forms,
+                        input_key,
+                        input_mfp,
+                    };
+                    insert_node(node_id, node);
+
+                    todo.push(*input);
+                }
+            };
+        }
+
+        // Apply replacements.
+        let replace = |id: &mut NodeId| {
+            // The replacement targets might have been replaced themselves, so iterate.
+            while let Some(target_id) = replacements.get(id) {
+                *id = *target_id;
+            }
+        };
+
+        for node in nodes.values_mut() {
+            for id in node.input_node_ids_mut() {
+                replace(id);
+            }
+        }
+        replace(&mut root);
+
+        flatten_order.reverse();
+        let topological_order = flatten_order;
+
+        Self {
+            nodes,
+            root,
+            topological_order,
+        }
+    }
+}
+
+impl<T> CollectionPlan for FlatPlan<T> {
+    fn depends_on_into(&self, out: &mut BTreeSet<GlobalId>) {
+        for node in self.nodes.values() {
+            if let FlatPlanNode::Get { id, .. } = node {
+                if let Id::Global(id) = id {
+                    out.insert(*id);
+                }
+            }
+        }
+    }
+}
+
+impl<T> FlatPlan<T> {
+    /// Return the ID of the root node.
+    pub fn root_id(&self) -> NodeId {
+        self.root
+    }
+
+    /// Return the root node.
+    fn root(&self) -> &FlatPlanNode<T> {
+        self.nodes.get(&self.root).expect("invariant (1)")
+    }
+
+    /// Return whether the plan contains recursion.
+    pub fn is_recursive(&self) -> bool {
+        // Because of invariant (4), every recursive plan must have a `LetRec` at its root.
+        matches!(self.root(), FlatPlanNode::LetRec { .. })
+    }
+
+    /// Split a recursive plan into its constituent (values, body) subplans.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this plan does not have a `LetRec` at the root.
+    pub fn split_recursive(
```

Comment:
This looks very similar to https://github.com/MaterializeInc/materialize/blob/0c43a2e49a767ba247edc8353537bdf481b23716/src/transform/src/analysis.rs#L105, but no need to jump to action!